File: frontend/src/api/construction.ts
/**
 * Construction Management API Client (Phase 2G)
 *
 * Provides type-safe access to:
 * - Contractor management
 * - Quality inspections
 * - Safety incidents
 * - Drawdown requests
 */

import { api } from './client'

// Enums matching backend
export type ContractorType =
  | 'general_contractor'
  | 'sub_contractor'
  | 'specialist'
  | 'consultant'
  | 'supplier'

export type InspectionStatus =
  | 'scheduled'
  | 'passed'
  | 'failed'
  | 'passed_with_conditions'
  | 'rectification_required'

export type SeverityLevel =
  | 'near_miss'
  | 'minor'
  | 'moderate'
  | 'severe'
  | 'fatal'

export type DrawdownStatus =
  | 'draft'
  | 'submitted'
  | 'verified_qs'
  | 'approved_architect'
  | 'approved_lender'
  | 'paid'
  | 'rejected'

// Response Types
export interface Contractor {
  id: string
  project_id: string
  company_name: string
  contractor_type: ContractorType
  contact_person: string | null
  email: string | null
  phone: string | null
  contract_value: number | null
  contract_date: string | null
  is_active: boolean
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string | null
}

export interface QualityInspection {
  id: string
  project_id: string
  development_phase_id: number | null
  inspection_date: string
  inspector_name: string
  location: string | null
  status: InspectionStatus
  defects_found: Record<string, unknown>
  photos_url: string[]
  notes: string | null
  created_at: string
}

export interface SafetyIncident {
  id: string
  project_id: string
  incident_date: string
  severity: SeverityLevel
  title: string
  description: string | null
  location: string | null
  reported_by: string | null
  is_resolved: boolean
  resolution_notes: string | null
  created_at: string
}

export interface DrawdownRequest {
  id: string
  project_id: string
  request_name: string
  request_date: string
  amount_requested: number
  amount_approved: number | null
  status: DrawdownStatus
  contractor_id: string | null
  supporting_docs: string[]
  notes: string | null
  created_at: string
  updated_at: string | null
}

// Create Types
export interface ContractorCreate {
  project_id: string
  company_name: string
  contractor_type?: ContractorType
  contact_person?: string
  email?: string
  phone?: string
  contract_value?: number
  contract_date?: string
  is_active?: boolean
  metadata?: Record<string, unknown>
}

export interface QualityInspectionCreate {
  project_id: string
  development_phase_id?: number
  inspection_date: string
  inspector_name: string
  location?: string
  status?: InspectionStatus
  defects_found?: Record<string, unknown>
  photos_url?: string[]
  notes?: string
}

export interface SafetyIncidentCreate {
  project_id: string
  incident_date: string
  severity: SeverityLevel
  title: string
  description?: string
  location?: string
  reported_by?: string
  is_resolved?: boolean
  resolution_notes?: string
}

export interface DrawdownRequestCreate {
  project_id: string
  request_name: string
  request_date: string
  amount_requested: number
  amount_approved?: number
  status?: DrawdownStatus
  contractor_id?: string
  supporting_docs?: string[]
  notes?: string
}

// Update Types
export interface ContractorUpdate {
  company_name?: string
  contractor_type?: ContractorType
  contact_person?: string
  email?: string
  phone?: string
  contract_value?: number
  contract_date?: string
  is_active?: boolean
  metadata?: Record<string, unknown>
}

export interface QualityInspectionUpdate {
  inspection_date?: string
  inspector_name?: string
  location?: string
  status?: InspectionStatus
  defects_found?: Record<string, unknown>
  photos_url?: string[]
  notes?: string
}

export interface SafetyIncidentUpdate {
  incident_date?: string
  severity?: SeverityLevel
  title?: string
  description?: string
  location?: string
  reported_by?: string
  is_resolved?: boolean
  resolution_notes?: string
}

export interface DrawdownRequestUpdate {
  request_name?: string
  request_date?: string
  amount_requested?: number
  amount_approved?: number
  status?: DrawdownStatus
  contractor_id?: string
  supporting_docs?: string[]
  notes?: string
}

// API Functions
export const constructionApi = {
  // Contractors
  async listContractors(
    projectId: string,
    type?: ContractorType,
  ): Promise<Contractor[]> {
    const params = type ? `?type=${type}` : ''
    return api.get<Contractor[]>(
      `/api/v1/projects/${projectId}/contractors${params}`,
    )
  },

  async createContractor(data: ContractorCreate): Promise<Contractor> {
    return api.post<Contractor>(
      `/api/v1/projects/${data.project_id}/contractors`,
      data,
    )
  },

  async updateContractor(
    contractorId: string,
    data: ContractorUpdate,
  ): Promise<Contractor> {
    return api.patch<Contractor>(
      `/api/v1/projects/contractors/${contractorId}`,
      data,
    )
  },

  // Quality Inspections
  async listInspections(
    projectId: string,
    status?: InspectionStatus,
  ): Promise<QualityInspection[]> {
    const params = status ? `?status=${status}` : ''
    return api.get<QualityInspection[]>(
      `/api/v1/projects/${projectId}/inspections${params}`,
    )
  },

  async createInspection(
    data: QualityInspectionCreate,
  ): Promise<QualityInspection> {
    return api.post<QualityInspection>(
      `/api/v1/projects/${data.project_id}/inspections`,
      data,
    )
  },

  async updateInspection(
    inspectionId: string,
    data: QualityInspectionUpdate,
  ): Promise<QualityInspection> {
    return api.patch<QualityInspection>(
      `/api/v1/projects/inspections/${inspectionId}`,
      data,
    )
  },

  // Safety Incidents
  async listIncidents(
    projectId: string,
    unresolvedOnly?: boolean,
  ): Promise<SafetyIncident[]> {
    const params = unresolvedOnly ? '?unresolved=true' : ''
    return api.get<SafetyIncident[]>(
      `/api/v1/projects/${projectId}/incidents${params}`,
    )
  },

  async createIncident(data: SafetyIncidentCreate): Promise<SafetyIncident> {
    return api.post<SafetyIncident>(
      `/api/v1/projects/${data.project_id}/incidents`,
      data,
    )
  },

  async updateIncident(
    incidentId: string,
    data: SafetyIncidentUpdate,
  ): Promise<SafetyIncident> {
    return api.patch<SafetyIncident>(
      `/api/v1/projects/incidents/${incidentId}`,
      data,
    )
  },

  // Drawdown Requests
  async listDrawdowns(
    projectId: string,
    status?: DrawdownStatus,
  ): Promise<DrawdownRequest[]> {
    const params = status ? `?status=${status}` : ''
    return api.get<DrawdownRequest[]>(
      `/api/v1/projects/${projectId}/drawdowns${params}`,
    )
  },

  async createDrawdown(data: DrawdownRequestCreate): Promise<DrawdownRequest> {
    return api.post<DrawdownRequest>(
      `/api/v1/projects/${data.project_id}/drawdowns`,
      data,
    )
  },

  async updateDrawdown(
    drawdownId: string,
    data: DrawdownRequestUpdate,
  ): Promise<DrawdownRequest> {
    return api.patch<DrawdownRequest>(
      `/api/v1/projects/drawdowns/${drawdownId}`,
      data,
    )
  },

  async approveDrawdown(
    drawdownId: string,
    approvedAmount?: number,
  ): Promise<DrawdownRequest> {
    const params = approvedAmount ? `?approved_amount=${approvedAmount}` : ''
    return api.post<DrawdownRequest>(
      `/api/v1/projects/drawdowns/${drawdownId}/approve${params}`,
      {},
    )
  },
}

// Status display helpers
export const contractorTypeLabels: Record<ContractorType, string> = {
  general_contractor: 'General Contractor',
  sub_contractor: 'Sub-Contractor',
  specialist: 'Specialist',
  consultant: 'Consultant',
  supplier: 'Supplier',
}

export const inspectionStatusLabels: Record<InspectionStatus, string> = {
  scheduled: 'Scheduled',
  passed: 'Passed',
  failed: 'Failed',
  passed_with_conditions: 'Passed with Conditions',
  rectification_required: 'Rectification Required',
}

export const severityLabels: Record<SeverityLevel, string> = {
  near_miss: 'Near Miss',
  minor: 'Minor',
  moderate: 'Moderate',
  severe: 'Severe',
  fatal: 'Fatal',
}

export const drawdownStatusLabels: Record<DrawdownStatus, string> = {
  draft: 'Draft',
  submitted: 'Submitted',
  verified_qs: 'QS Verified',
  approved_architect: 'Architect Approved',
  approved_lender: 'Lender Approved',
  paid: 'Paid',
  rejected: 'Rejected',
}

export const inspectionStatusColors: Record<
  InspectionStatus,
  'default' | 'success' | 'error' | 'warning' | 'info'
> = {
  scheduled: 'info',
  passed: 'success',
  failed: 'error',
  passed_with_conditions: 'warning',
  rectification_required: 'error',
}

export const severityColors: Record<
  SeverityLevel,
  'default' | 'success' | 'error' | 'warning' | 'info'
> = {
  near_miss: 'info',
  minor: 'default',
  moderate: 'warning',
  severe: 'error',
  fatal: 'error',
}

export const drawdownStatusColors: Record<
  DrawdownStatus,
  'default' | 'success' | 'error' | 'warning' | 'info'
> = {
  draft: 'default',
  submitted: 'info',
  verified_qs: 'info',
  approved_architect: 'warning',
  approved_lender: 'success',
  paid: 'success',
  rejected: 'error',
}

File: backend/app/schemas/construction.py
"""Pydantic schemas for Phase 2G Construction Delivery Module."""

from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import Any, List
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field

from app.models.construction import (
    ContractorType,
    InspectionStatus,
    SeverityLevel,
    DrawdownStatus,
)


class ContractorBase(BaseModel):
    company_name: str
    contractor_type: ContractorType = ContractorType.GENERAL_CONTRACTOR
    contact_person: str | None = None
    email: str | None = None
    phone: str | None = None
    contract_value: Decimal | None = None
    contract_date: date | None = None
    is_active: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)


class ContractorCreate(ContractorBase):
    project_id: UUID


class ContractorUpdate(BaseModel):
    company_name: str | None = None
    contractor_type: ContractorType | None = None
    contact_person: str | None = None
    email: str | None = None
    phone: str | None = None
    contract_value: Decimal | None = None
    contract_date: date | None = None
    is_active: bool | None = None
    metadata: dict[str, Any] | None = None


class ContractorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    company_name: str
    contractor_type: ContractorType = ContractorType.GENERAL_CONTRACTOR
    contact_person: str | None = None
    email: str | None = None
    phone: str | None = None
    contract_value: Decimal | None = None
    contract_date: date | None = None
    is_active: bool = True
    metadata: dict[str, Any] = Field(
        default_factory=dict, validation_alias="metadata_json"
    )
    created_at: datetime
    updated_at: datetime | None = None


# Inspections
class QualityInspectionBase(BaseModel):
    inspection_date: date
    inspector_name: str
    location: str | None = None
    status: InspectionStatus = InspectionStatus.SCHEDULED
    defects_found: dict[str, Any] = Field(default_factory=dict)
    photos_url: List[str] = Field(default_factory=list)
    notes: str | None = None


class QualityInspectionCreate(QualityInspectionBase):
    project_id: UUID
    development_phase_id: int | None = None


class QualityInspectionUpdate(BaseModel):
    inspection_date: date | None = None
    inspector_name: str | None = None
    location: str | None = None
    status: InspectionStatus | None = None
    defects_found: dict[str, Any] | None = None
    photos_url: List[str] | None = None
    notes: str | None = None


class QualityInspectionResponse(QualityInspectionBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    development_phase_id: int | None = None
    created_at: datetime


# Safety
class SafetyIncidentBase(BaseModel):
    incident_date: datetime
    severity: SeverityLevel
    title: str
    description: str | None = None
    location: str | None = None
    reported_by: str | None = None
    is_resolved: bool = False
    resolution_notes: str | None = None


class SafetyIncidentCreate(SafetyIncidentBase):
    project_id: UUID


class SafetyIncidentUpdate(BaseModel):
    incident_date: datetime | None = None
    severity: SeverityLevel | None = None
    title: str | None = None
    description: str | None = None
    location: str | None = None
    reported_by: str | None = None
    is_resolved: bool | None = None
    resolution_notes: str | None = None


class SafetyIncidentResponse(SafetyIncidentBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    created_at: datetime


# Drawdowns
class DrawdownRequestBase(BaseModel):
    request_name: str
    request_date: date
    amount_requested: Decimal
    amount_approved: Decimal | None = None
    status: DrawdownStatus = DrawdownStatus.DRAFT
    contractor_id: UUID | None = None
    supporting_docs: List[str] = Field(default_factory=list)
    notes: str | None = None


class DrawdownRequestCreate(DrawdownRequestBase):
    project_id: UUID


class DrawdownRequestUpdate(BaseModel):
    request_name: str | None = None
    request_date: date | None = None
    amount_requested: Decimal | None = None
    amount_approved: Decimal | None = None
    status: DrawdownStatus | None = None
    contractor_id: UUID | None = None
    supporting_docs: List[str] | None = None
    notes: str | None = None


class DrawdownRequestResponse(DrawdownRequestBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    created_at: datetime
    updated_at: datetime | None = None
